File: frontend/services/chatbot.ts
/**
 * Chatbot API service — wired to backend POST /chatbot/chat (persists conversation + messages).
 */
import { apiClient } from './api';

export interface ChatWithHistoryRequest {
  patient_id: string;
  doctor_query: string;
  mode?: string;
  conversation_id?: string | null;
  subject?: string;
  start_new?: boolean;
}

export interface ChatWithHistoryResponse {
  response: {
    final_message: string;
    detailed_message?: string | null;
    summary_message?: string | null;
    agent_outputs?: Record<string, unknown>;
    mode?: string;
    patient_id?: string;
    report_created?: boolean;
    failed?: boolean;
  };
  conversation_id: string;
  message_id_user?: string;
  message_id_assistant?: string;
  transcript?: { id?: string; role: string; content: string; created_at?: string }[];
}

export interface ConversationTranscriptResponse {
  conversation_id: string | null;
  patient_id: string;
  transcript: { id?: string; role: string; content: string; created_at?: string }[];
}

export interface ConversationSummary {
  conversation_id: string;
  patient_id: string;
  subject: string;
  preview: string;
  message_count: number;
  created_at?: string;
  updated_at?: string;
}

export interface PatientConversationListResponse {
  patient_id: string;
  conversations: ConversationSummary[];
}

class ChatbotService {
  /**
   * Send a message and get assistant reply. Pass conversation_id to continue a thread.
   */
  async chat(payload: ChatWithHistoryRequest): Promise<ChatWithHistoryResponse> {
    const body: Record<string, unknown> = {
      patient_id: payload.patient_id,
      doctor_query: payload.doctor_query,
      mode: payload.mode ?? 'master',
      subject: payload.subject ?? 'Clinical chat',
    };
    if (payload.conversation_id) {
      body.conversation_id = payload.conversation_id;
    }
    if (payload.start_new) {
      body.start_new = true;
    }
    return await apiClient.post<ChatWithHistoryResponse>('/chatbot/chat', body);
  }

  /**
   * Get patient context (risk, summary, recommendations) when opening chat for a patient.
   */
  async getPatientContext(patientId: string): Promise<Record<string, unknown>> {
    return await apiClient.get<Record<string, unknown>>(`/chatbot/patient-context/${patientId}`);
  }

  async getLatestConversation(patientId: string): Promise<ConversationTranscriptResponse> {
    return await apiClient.get<ConversationTranscriptResponse>(
      `/chatbot/patients/${patientId}/conversation`,
    );
  }

  async listPatientConversations(patientId: string): Promise<PatientConversationListResponse> {
    return await apiClient.get<PatientConversationListResponse>(
      `/chatbot/patients/${patientId}/conversations`,
    );
  }

  async getConversationTranscript(
    conversationId: string,
    patientId?: string,
  ): Promise<ConversationTranscriptResponse> {
    const query = patientId ? `?patient_id=${encodeURIComponent(patientId)}` : '';
    return await apiClient.get<ConversationTranscriptResponse>(
      `/chatbot/conversations/${conversationId}/transcript${query}`,
    );
  }
}

export const chatbotService = new ChatbotService();

File: backend/app/schemas/chatbot.py
"""Chatbot request and response schemas for the clinical assistant."""
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


class ChatMessage(BaseModel):
    """Single message in a conversation (for multi-turn chat)."""
    role: str = Field(..., description="user | assistant")
    content: str = Field(default="", description="Message text")


class DoctorAssessment(BaseModel):
    """Optional doctor input for compare mode."""
    diagnosis: Optional[str] = None
    planned_medications: List[str] = Field(default_factory=list)
    planned_lifestyle: List[str] = Field(default_factory=list)


class WhatIfChanges(BaseModel):
    """Hypothetical changes for what-if scenario (e.g. BMI, HbA1c)."""
    # Allow arbitrary numeric/string overrides for context fields
    model_config = {"extra": "allow"}


class ChatbotRequest(BaseModel):
    """Request to the clinical chatbot. Supports single-turn and multi-turn chat."""
    patient_id: str
    doctor_query: str = Field(default="", description="Free-text query from clinician (current message)")
    mode: str = Field(
        default="recommend",
        description="recommend | explain | what_if | compare | auto | master (master infers and calls agents/graph explainer as tools)"
    )
    message_history: Optional[List[ChatMessage]] = Field(
        default=None,
        description="Previous messages in this conversation (user/assistant); used for context and audit"
    )
    doctor_assessment: Optional[DoctorAssessment] = None
    what_if_changes: Optional[Dict[str, Any]] = None  # e.g. {"BMI": 27.0, "HbA1c": 7.1}


class ChatbotResponse(BaseModel):
    """Response from the clinical chatbot."""
    mode: str
    patient_id: str
    agent_outputs: Dict[str, Any] = Field(
        default_factory=dict,
        description="risk, lifestyle, medication, explainability, comparison, whatif as applicable"
    )
    final_message: str
    detailed_message: Optional[str] = Field(
        default=None,
        description="Expanded clinician-facing response when lifestyle or medication guidance is returned"
    )
    summary_message: Optional[str] = Field(
        default=None,
        description="Optional compact summary retained for debugging or compact-card use"
    )
    doctor_note: str = Field(default="The doctor remains the final decision-maker.")


class ChatWithHistoryRequest(BaseModel):
    """Request for chat endpoint that persists conversation and messages (ChatGPT-style)."""
    patient_id: str
    doctor_query: str = Field(..., description="Current user message")
    mode: str = Field(default="recommend", description="recommend | explain | what_if | compare | auto | master")
    conversation_id: Optional[str] = Field(None, description="Existing conversation; if omitted, a new one is created")
    subject: Optional[str] = Field(None, description="Subject for new conversation (default: Clinical chat)")
    doctor_assessment: Optional[DoctorAssessment] = None
    what_if_changes: Optional[Dict[str, Any]] = None


class TranscriptMessage(BaseModel):
    """Single message in transcript (for API response)."""
    id: Optional[str] = None
    role: str = Field(..., description="user | assistant")
    content: str = ""
    created_at: Optional[str] = None


class ConversationSummary(BaseModel):
    """Conversation list item for chatbot session picker."""
    conversation_id: str
    patient_id: str
    subject: str = "Clinical chat"
    preview: str = ""
    message_count: int = 0
    created_at: Optional[str] = None
    updated_at: Optional[str] = None


class PatientConversationListResponse(BaseModel):
    """All conversations for one patient, newest first."""
    patient_id: str
    conversations: List[ConversationSummary] = Field(default_factory=list)


class ConversationTranscriptResponse(BaseModel):
    """Transcript response for one conversation or latest conversation."""
    conversation_id: Optional[str] = None
    patient_id: str
    transcript: List[TranscriptMessage] = Field(default_factory=list)


class ChatWithHistoryResponse(BaseModel):
    """Response from chat endpoint: chatbot reply + conversation id + full transcript."""
    response: ChatbotResponse
    conversation_id: str
    message_id_user: Optional[str] = None
    message_id_assistant: Optional[str] = None
    transcript: List[TranscriptMessage] = Field(default_factory=list)
